### PR TITLE
Woocommerce Integration

### PR DESCRIPTION
**Pull request by vinayak--**
# WooCommerce Integration 

***Setting Up WooCommerce on ERPNEXT:-***

Steps:- 

1. From Awesome-bar, go to "Woocommerce Settings" doctype.

2. From your woocommerce site, generate "API consumer key" and  "API consumer secret" using Keys/Apps tab.

3. Paste those generated "API consumer key" and  "API consumer secret" into "Woocommerce Settings" doctype.

4. In "Woocommerce Server URL" paste the url of your site where ERPNEXT is installed.

5. Make sure "Enable Sync" is checked.

6. Select Account type from Account Details Section.

7. Click Save.

8. After saving, "Secret" and "Endpoint" are generated automatically and can be seen on "Woocommerce Settings" doctype.

9. Now from your woocommerce site, click on webhooks option and click on "Add Webhook".

10. Give name to the webhook of your choice. Click on Status dropdown and select "Active". Select Topic as "Order Created". Copy the "Endpoint" from "Woocommerce Settings" doctype and paste it in "Delivery URL" field. Copy "Secret" from "Woocommerce Settings" doctype and paste it in "Secret" field. Keep API VERSION as it is and click on Save Webhook.

11. Now the WooCommerce is successful setup on your system.

Woocommerce Integration

### Note:-  In above gif, inplace of delivery url on woocommerce website, you need to paste the url you will obtain after saving the "Woocommerce Settings" page (i.e. Endpoint from "Woocommerce Settings"). I pasted other url because I was using localhost. Please paste your endpoint in place of Delivery URL.
	

***WooCommerce Integration Working:-*** 

Steps:- 

1. From your Woocommerce website, register yourself as a user.

2. Now Click on Address Details and provide the required details.

3. For start shopping, click on Shop option and now available products can be seen.

4. Add the desired products into cart and click on View Cart.

5. From Cart, once you have added the desired products, you can click on proceed to checkout.

6. All billing details and Order details can be seen now. Once you are ok with it, click on Place Order button.

7. "Order Received" message can been seen indicating that the order is placed successfully.

8. Now on system where ERPNEXT is installed check the following doctypes: Customer, Address, Item, Sales Order.

Woocommerce Integration

https://github.com/frappe/erpnext/issues/12873

- [x] Gif / Screenshots / Videos for PR
- [x] Docs
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by revant** [comment]
We need to add `Woocommerce Settings` a module page. Right now the only way to access it is awesome bar.

File: erpnext/tests/test_woocommerce.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		woo_settings.enable_sync = 1
+		woo_settings.tax_account = "Sales Expenses - W"
+		woo_settings.f_n_f_account = "Expenses - W"
+		woo_settings.save(ignore_permissions=True)
+
+		frappe.db.commit()
+
+
+		r = emulate_request()
+		print(r.text)
```
Comment (revant) [inline erpnext/tests/test_woocommerce.py]: is this necessary? keep it for local test for debugging.

File: erpnext/tests/test_woocommerce.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		self.assertTrue(r.status_code == 200)
+		self.assertTrue(frappe.get_value("Customer",{"woocommerce_email":"tony@gmail.com"}))
+		self.assertTrue(frappe.get_value("Item",{"woocommerce_id": 56}))
+		self.assertTrue(frappe.get_value("Sales Order",{"woocommerce_id":74}))
+
+		# cancel & delete order
```
Comment (revant) [inline erpnext/tests/test_woocommerce.py]: also remove commented code.

Add one more test for repeat orders

**Comment by Olawale1** [comment]
Hi @vinayak-- 

Awesome job! How does this work when the customer makes payment directly on the woocommerce site (i.e using Credit Card etc)? 

Cheers

**Comment by vinayak--** [comment]
@Olawale1 
Hey! 
Right now we don't have that functionality of payment via CC etc. But we will surely think of  implementing it in next phases. Till then try testing the current product.
Thanks.

**Comment by kothagunda** [comment]
Very userful feature guys. 
Since we are using both WooCommerce and ERPNExt in an un-integrated way right now, it will definitely help us saving our time and gives a good user experience as the stock balances will be real time. 

I have below queries, pls clarify them:
1. What about the existing users we already have on wooCommerce site (that we dont have on erpnext), do they get automatically added on my erpnext server in the customer List after the initial setup. 
2. Currently i have products, and prices already on woocommerce site. Should i first delete all those. So that my item list from the erpnext server starts showing up in woocommerce. (we have totally different item codes on woocommerce and erpnext sites)
3. After the sales order is placed in woocommerce. Do i have to create a delivery note or a sales invoice from erpnext . If so, do these get synched, and shown to user on woocommerce site.

Thank you

**Comment by Suprazz** [comment]
Sorry for my ignorance but how can we test that? I tried on the develop branch with a vm and cannot see anything related with woo commerce. Thanks

**Comment by Suprazz** [comment]
This is great! I was finally able to test it and make it work. Item are automatically created in ERPNext from WooCommerce. How do we do if we already have existing items to link (insteat of creating new one).

Do you think there is a way to sync inventory too eventually.

I'm ready to pay for the development of this. Thanks

**Comment by vinayak--** [comment]
@albertusgeyser
Hey!!
Please switch to develop branch on ERPNext as well. Now you should be able to see the integration.
Thanks!

**Comment by vinayak--** [comment]
@kothagunda
Hey!!
You need not delete your current products on Woocommerce site. All new users/products or existing users/products will be mapped to the system whenever a new order comes.
We are already thinking about syncing the past orders or products into the system directly once the setup is done. And as far as payment option is concerned we are also thinking to implement it into next phase.
Thanks!!

**Comment by decryptedchaos** [comment]
Not sure if this is the proper place to report issues,  but it seems better than opening a direct issue to erpnext.  

The callback webhook is returning a 500 error,  is this a known issue? 

I can call it from curl cmd line and get success but when woocommerce calls it it gets back 500.  


 ```
[Webhook Delivery] => Array
        (
            [Delivery ID] => 9f4224d8129dfb717cd08be19b5d13f4
            [Date] => Apr 18, 2018 @ 4:00
            [URL] => http://erp.*********.tk/api/method/erpnext.erpnext_integrations.connectors.woocommerce_connection.order
            [Duration] => 1.11762
            [Request] => Array
                (
                    [Method] => POST
                    [Headers] => Array
                        (
                            [User-Agent] => WooCommerce/3.3.5 Hookshot (WordPress/4.9.5)
                            [Content-Type] => application/json
                            [X-WC-Webhook-Source] => https://************.tk/
                            [X-WC-Webhook-Topic] => order.created
                            [X-WC-Webhook-Resource] => order
                            [X-WC-Webhook-Event] => created
                            [X-WC-Webhook-Signature] => RNwwg9jBOZvV66JbS+GzCtU760qU6m6NlAH79qSVhR4=
                            [X-WC-Webhook-ID] => 1
                            [X-WC-Webhook-Delivery-ID] => 9f4224d8129dfb717cd08be19b5d13f4
                        )

                )

        
            [Response] => Array
                (
                    [Code] => 500
                    [Message] => Internal Server Error
                    [Headers] => Requests_Utility_CaseInsensitiveDictionary Object
                        (
                            [data:protected] => Array
                                (
                                    [content-type] => text/html; charset=utf-8
                                    [content-length] => 7194
                                    [x-page-name] => message
                                    [x-from-cache] => False
                                    [server] => Werkzeug/0.14.1 Python/3.5.2
                                    [date] => Wed, 18 Apr 2018 04:00:39 GMT
                                )

                        )
```

**Comment by kothagunda** [comment]
Thanks @vinayak-- for your response.

**Comment by albertusgeyser** [comment]
When should it be available in Production?

**Comment by dolispy** [comment]
Hi I just followed your setup steps for the woocommerce integration now and I am not getting any updates in the doctypes.
I am also a bit confused about the woocommerce url...you wrote paste the url of the site where ERPnext is installed but on your demo you used the url of your woocommerce site. Is there something I am missing? because I have used both urls and still no updates in any of the customer, item, sales order doctypes.

**Comment by lutealit** [comment]
Hi. Im interesting on it and can be a tester. How can I try erpnext woocommerce_integration?

Thanks

**Comment by albertusgeyser** [comment]
Not sure why?, but cant also not see integration on development.
Could anyone guide us not seeing it how to see integration?

**Comment by nikitogr** [comment]
Successfully connected but it doesn't seem to be updated in erpnext. The order is being made without the user singing in, just filling in his details before the order after the cart. Is the integration based on the user account?

**Comment by nikitogr** [comment]
Hello,

Getting 403 permission errors. Only customer first & last name is updated and imported. Nothing else, while on 403 error. The fix to even just get the customer name is to set guest role permissions. But only in customers working.

I havent found any solution on 403 permission errors......

**Comment by Mata101** [comment]
This integrations is a great help but there are things that I want to clear.
I am planning to use the current woocoommerce to erpnext integration but there are things that I think it's a problem during my tests.
In placing an order in woocommerce shop, the customer can put its company name and when erpnext receives the request through webhooks in woocommerce it fails with an error. Then I found out the cause of the error, the company that the customer inputted is expected to be already existing in erpnext side. And when the customer won't input any company, the process of the request will use the default company in erpnext. Do the devs expected the buyers to be single persons only without company, so they can use the default company? Haven't they considered that the customer might be a company? and that this company do not exist in erpnext? or do they expect only companies that were previously created in erpnext?
I also tried to create the company in erpnext side first before I place an order, to assume that the customer will input a company that is already existing in erpnext. But it still gets an error because the accounting details that is already set in Woocommerce settings is from another company (I am talking about the accounts that is used in accounting details), which means these accounts belong to only one company, so if different companies are found for every request when an order is placed, it will still get an error because the accounts that are selected in the accounting details inside Woocommerce Settings is statically set from one company only. Why do they allow the users to input the accounting details inside woocommerce settings even though they are expecting different companies from different orders in woocommerce?
Can someone please answer this questions to make this things clear?

**Comment by taherkhalil** [comment]
Hello @vinayak--  , 
when is the planned next phase of the app ?

the main things required to use wocommerce with erpnext on live transcations right now are following according to my understanding. 
1.Two way sync of items -already present items on Erpnext should show on Woocomerce ( something like the shopify thing - sync to wocomerce check ) -  
.
2. The payment finalization i.e - if the invoice is paid online through woocomerce , it should update in erpnext system as paid . and make necessary JE/PE entries. 

3. Stock maintenance i.e out of stock items should reflect on wocommerece.

Willing to fund this.
please inform .

**Comment by germanllop** [comment]
Hello! Any updates on this integration?

**Comment by PiotrCzapla** [comment]
I've found this  integration that does the product sync as well: https://github.com/libracore/WooCommerceConnector

From the reading of the source code it seems to be doing what one would expect from a proper woocomerce integaration:
- sync products with variants and different tax periodically
- keep the quantity up to date
- sync orders from woocommmerce

I'm going to test it soon with ERPNext v12

**Comment by Suprazz** [comment]
there is an issue with v12. see here https://discuss.erpnext.com/t/woocommerce-full-2-way-integration/42432/42

**Comment by wrentedspace** [comment]
Hi @vinayak-- and @revant ; I am loving the sync from WooCommerce to ERPNext. However, some customers have a shipping address and a separate billing address, and only the billing address is getting imported through the integration. How can we get the shipping address to be added to the customer record automatically? Thanks!

**Comment by olamide2** [comment]
> I've found this integration that does the product sync as well: https://github.com/libracore/WooCommerceConnector
> 
> From the reading of the source code it seems to be doing what one would expect from a proper woocomerce integaration:
> 
> * sync products with variants and different tax periodically
> * keep the quantity up to date
> * sync orders from woocommmerce
> 
> I'm going to test it soon with ERPNext v12

Did this work as expected?

**Comment by PiotrCzapla** [comment]
nope, ERPNext v12 was way to buggy for me to get it to work before i could test it so I decided to wait.

**Pull request merged**